### PR TITLE
Add support for WCF web services using SOAP 1.2

### PR DESCRIPTION
**Pull request by jonsabados**
Hello - this pull request would add support for sending a Content-Type header with a value along the lines of 'application/soap+xml; charset=utf-8; action="http://foo.bar.com/blah/dosomething"' for SOAP 1.2 services (specifically WCF based services but should work for others).

Noteworthy bits: 
* I couldn't figure a way to do this without changing the RoundTrip method signature which would break any previously generated code so I added an additional RoundTripSoap12 method & am calling that when there is a binding.operation.operation element bound to http://schemas.xmlsoap.org/wsdl/soap12/ with a soapAction attribute present - more than happy to change this if there's a better way of figuring out when to send the soap 1.2 style content-type header (or if there are any other things that should be cleaned up). 
* There was a failing unit test present in the area of the code I was going to be modifying due to a test sending in nil input & the reflection based setting of the SOAPAction header causing a panic  - I wrapped that guy in a nil check so that I'm not submitting a PR with broken tests but could rip that change out if its appropriate, or see about modifying the test to trigger a failure in a different manner.  Avoided doing so though as I was not sure if the intent of the test was to test nil input or if that was just a way of throwing a wrench at things to get a failure.
* I thought about doing something similar to the way the SAOPAction header is set for the action on the soap 1.2 requests but since it required a different method call anyways and that information was specifically present in WSDLs it felt cleaner to avoid reflection & dump in the action directly.

### CODE REVIEW

File: soap/client.go [hunk 1]
```diff
@@ -96,6 +91,31 @@ func (c *Client) RoundTrip(in, out Message) error {
...
 
+// RoundTrip implements the RoundTripper interface.
+func (c *Client) RoundTrip(in, out Message) error {
+	headerFunc := func(r *http.Request) {
+		ct := c.ContentType
+		if ct == "" {
+			ct = "text/xml"
+		}
+		r.Header.Set("Content-Type", ct)
+		// TODO - what should happen here? there is a test case that does send in a nil & it was triggering a panic,
```
Comment (fiorix) [inline soap/client.go]: How was this not failing before?

Well, worst case scenario is that we cannot set a SOAPAction and the request may fail - this probably depends on the server tho, and there's all kinds of them out there. I believe for now we can keep the check and just not set the SOAPAction.

File: wsdlgo/encoder.go [hunk 2]
```diff
@@ -499,6 +499,21 @@ var soapFuncT = template.Must(template.New("soapFunc").Parse(
 }
 `))
 
+var soapFuncT12 = template.Must(template.New("soapFunc12").Parse(
```
Comment (fiorix) [inline wsdlgo/encoder.go]: Don't want to be picky, but why not soap12FuncT ?

File: wsdlgo/encoder.go [hunk 3]
```diff
@@ -515,28 +530,57 @@ func (ge *goEncoder) writeSOAPFunc(w io.Writer, d *wsdl.Definitions, op *wsdl.Op
...
+			strings.ToLower(d.PortType.Name[:1]) + d.PortType.Name[1:],
+			strings.Title(op.Name),
+			strings.Join(in, ","),
+			strings.Join(out, ","),
+			strings.TrimPrefix(typ[1], "*"),
+			strings.TrimPrefix(xmlToken, "*"),
+			typ[1][0] == '*',
+			ret[0],
+		})
+	} else {
```
Comment (fiorix) [inline wsdlgo/encoder.go]: Could probably drop the else and return true here, might read better.

File: soap/client.go (hunk 1 above)
Comment (jonsabados) [inline soap/client.go]: I think it has been (at least it does locally on revision 40fd89a0a40377adf9b1baaf24a37fa7fc67be45) - regardless left the check in & removed the comment

File: wsdlgo/encoder.go (hunk 2 above)
Comment (jonsabados) [inline wsdlgo/encoder.go]: that actually is way better naming - updated

File: wsdlgo/encoder.go (hunk 3 above)
Comment (jonsabados) [inline wsdlgo/encoder.go]: Makes sense, updated

**Comment by fiorix** [comment]
👍

**Pull request merged**